### DISCUSSION TITLE
Integrating the visual builder into third-party applications

### DISCUSSION THREAD

**Discussion by yashwason** [webstudio-is/webstudio-community#93] [Q&A] [answered]
I stumbled upon WebStudio today, and I am in love! The quality of what you guys are building is outstanding.

We are building an application that requires a Visual Builder and the builder needs to be integrated with our custom CMS APIs. I am wondering if WebStudio is a viable option for this purpose as it would reduce our development time by months.

I understand WebStudio is extensible and can be integrated with third-party APIs, but we need the Visual Builder to be **part** of our application - as in be accessible to users after their login to our app. A good example would be Shopify and how they provide a Visual Builder to their users - although our use-case is not e-commerce but the requirements are exactly the same as Shopify's builder module.

The application that the Visual Builder will be part of will be built in NextJs (maybe Remix).

Is this something WebStudio can handle? If yes, is it possible in the Open Source offering or we should reach out for a paid plan? Any guidance would be helpful.

**Comment by kof** [comment]
>  builder needs to be integrated with our custom CMS APIs

- yeah we are working towrads launching datasources that allow to connect to any endpoint, this is launching in weeks
- then we are working towards first cms integration

> it would reduce our development time by months.

Months? So you estimate you could build webstudio alternative in months? 🤣 Its years, my friend.

>   but we need the Visual Builder to be part of our application

AGPL  is designed for free usage when the user doesn't modify or open sources the things they modify.

To understand exactly what you would need it will take more information and probably a conversation.

Generally webstudio cloud service is more than the open core, so it depends on what you actually need https://webstudio.is/pricing

We do plan to offer enterprise and white-label licenses for companies who want to leverage the full product with all current and upcoming features that are designed around needs for teams. Our goal is to keep all features an individual needs for free and open source, while keeping features that companies need as not-free / closed source. This is required to avoid corporate cannibalism.

**Reply by yashwason** [reply]
Thanks for the quick response. This makes sense.

Although being able to edit the source is important to us as we plan to add features to the editor. So I guess WebStudio would not be the right choice for us.

**Reply by kof** [reply]
Webstudio is actually generating code that we will make properly readable in the near future (aka compiler). That code shouldn't be modified otherwise you diverge between visual and code

**Reply by yashwason** [reply]
I'm not sure I get that. I understand the editor generates code.

But you meant that the WebStudio codebase/source code itself cannot be modified right?

**Reply by kof** [reply, marked as answer]
if you generate code from webstudio (code of your site or components), then modify that code, then you won't be seeing the same thing in the builder as you do in the app where you run the modified that code.

But I am realizing you meant editing webstudio builder itself.

In that case:
1. how you edit it is important, we are building it with extensibility in mind, simply modifying random stuff will get you in trouble anyways, because you want to be able to upgrade to the latest versions
2. you can do it with webstudio, but it will require you to buy a license, since this is literally extending/modifying it and reselling it.

**Reply by kof** [reply]
Webstudio is extensible in 2 levels:
1. literally swapping out certain packages with stable interfaces
2. using our public apis for integrations which will become public soon.

**Reply by kof** [reply]
Potentially even a 3rd level - using webstudio builder UI as a library - customizing the behavior globally, though this is right now not a priority, at least not if there are significant customizations needed.
